File: Graph/Disjoint-Set/dis.cpp
#include <vector>
using namespace std;

class Solution {
  public:
    int findParent(int a,vector<int>& parent){
        if(parent[a] == a)return a;
        return parent[a] = findParent(parent[a],parent);
    }
    void unionGroup(int a, int b, vector<int>& parent){
        int parB = findParent(b,parent);
        int parA = findParent(a,parent);
        parent[parA] = parB;
        return;
    }
    vector<int> DSU(int n, vector<vector<int>>& queries) {
        // code here
        vector<int>parent(n+1,0);
        for(int i=0;i<=n;i++){
            parent[i] = i;
        }
        vector<int>ans;
        int querySize = queries.size();
        for(int i=0;i<querySize;i++){
            if(queries[i][0]==2)ans.push_back(findParent(queries[i][1],parent));
            else unionGroup(queries[i][1],queries[i][2],parent);
        }
        return ans;
    }
};

File: Graph/Disjoint-Set/dis.py
class Solution:
    def findParent(self, a, parent):
        if parent[a] == a:
            return a
        parent[a] = self.findParent(parent[a], parent)
        return parent[a]

    def unionGroup(self, a, b, parent):
        parA = self.findParent(a, parent)
        parB = self.findParent(b, parent)
        parent[parA] = parB

    def DSU(self, n, queries):
        parent = [i for i in range(n + 1)]
        ans = []

        for query in queries:
            if query[0] == 2:
                ans.append(self.findParent(query[1], parent))
            else:
                self.unionGroup(query[1], query[2], parent)

        return ans